File: export-data.js
import { neon } from "@neondatabase/serverless";
import fs from "fs";

const sql = neon(process.env.DATABASE_URL);

async function exportData() {
  console.log("Starting data export...");
  
  let output = `-- Vespro Database Data Export
-- Generated for n8n Agent Integration
-- Database: vespro schema

SET search_path TO vespro;

`;

  try {
    // Export forms table
    console.log("Exporting forms table...");
    const forms = await sql`SELECT * FROM vespro.forms ORDER BY created_at`;
    
    if (forms.length > 0) {
      output += "-- Forms table data\n";
      for (const form of forms) {
        const columns = Object.keys(form).join(", ");
        const values = Object.entries(form).map(([key, val]) => {
          if (val === null) return "NULL";
          if (key === 'calculated_values' || key === 'metadata' || key === 'extra') {
            return `'${JSON.stringify(val)}'::jsonb`;
          }
          if (typeof val === 'boolean') return val ? 'TRUE' : 'FALSE';
          if (typeof val === 'number') return val;
          if (val instanceof Date) return `'${val.toISOString()}'`;
          // Escape single quotes in strings
          const escaped = String(val).replace(/'/g, "''");
          return `'${escaped}'`;
        }).join(", ");
        
        output += `INSERT INTO vespro.forms (${columns}) VALUES (${values});\n`;
      }
      console.log(`Forms exported: ${forms.length} records`);
    }

    // Export cost_items table
    console.log("Exporting cost_items table...");
    const costItems = await sql`SELECT * FROM vespro.cost_items ORDER BY item_id`;
    
    if (costItems.length > 0) {
      output += "\n-- Cost_items table data\n";
      for (const item of costItems) {
        const columns = Object.keys(item).join(", ");
        const values = Object.entries(item).map(([key, val]) => {
          if (val === null) return "NULL";
          if (key === 'extra') {
            return `'${JSON.stringify(val)}'::jsonb`;
          }
          if (typeof val === 'boolean') return val ? 'TRUE' : 'FALSE';
          if (typeof val === 'number') return val;
          if (val instanceof Date) return `'${val.toISOString()}'`;
          // Escape single quotes in strings
          const escaped = String(val).replace(/'/g, "''");
          return `'${escaped}'`;
        }).join(", ");
        
        output += `INSERT INTO vespro.cost_items (${columns}) VALUES (${values});\n`;
      }
      console.log(`Cost_items exported: ${costItems.length} records`);
    }

    // Export materials table
    console.log("Exporting materials table...");
    const materials = await sql`SELECT * FROM vespro.materials`;
    
    if (materials.length > 0) {
      output += "\n-- Materials table data\n";
      for (const material of materials) {
        const columns = Object.keys(material).join(", ");
        const values = Object.entries(material).map(([key, val]) => {
          if (val === null) return "NULL";
          if (typeof val === 'number') return val;
          // Escape single quotes in strings
          const escaped = String(val).replace(/'/g, "''");
          return `'${escaped}'`;
        }).join(", ");
        
        output += `INSERT INTO vespro.materials (${columns}) VALUES (${values});\n`;
      }
      console.log(`Materials exported: ${materials.length} records`);
    }

    // Export cost_groups table
    console.log("Exporting cost_groups table...");
    const costGroups = await sql`SELECT * FROM vespro.cost_groups`;
    
    if (costGroups.length > 0) {
      output += "\n-- Cost_groups table data\n";
      for (const group of costGroups) {
        const columns = Object.keys(group).join(", ");
        const values = Object.entries(group).map(([key, val]) => {
          if (val === null) return "NULL";
          if (typeof val === 'number') return val;
          // Escape single quotes in strings
          const escaped = String(val).replace(/'/g, "''");
          return `'${escaped}'`;
        }).join(", ");
        
        output += `INSERT INTO vespro.cost_groups (${columns}) VALUES (${values});\n`;
      }
      console.log(`Cost_groups exported: ${costGroups.length} records`);
    }

    // Write to file
    fs.writeFileSync('insertdata.sql', output);
    
    const stats = fs.statSync('insertdata.sql');
    console.log("\n✅ Export completed successfully!");
    console.log(`📁 File: insertdata.sql`);
    console.log(`📏 Size: ${(stats.size / 1024).toFixed(2)} KB`);
    console.log(`📝 Lines: ${output.split('\n').length}`);
    
  } catch (error) {
    console.error("Export failed:", error);
    process.exit(1);
  }
}

exportData();

File: export_data.py
import os
import psycopg2
from psycopg2.extras import RealDictCursor
import json

# Database connection from environment
DATABASE_URL = os.environ.get('DATABASE_URL')

def escape_string(s):
    if s is None:
        return 'NULL'
    if isinstance(s, (dict, list)):
        s = json.dumps(s)
    s = str(s).replace("'", "''")
    return f"'{s}'"

def export_to_sql():
    conn = psycopg2.connect(DATABASE_URL)
    cur = conn.cursor(cursor_factory=RealDictCursor)
    
    with open('insertdata.sql', 'w') as f:
        f.write('-- Vespro Database Data Export\n')
        f.write('-- Generated for n8n Agent Integration\n\n')
        f.write('SET search_path TO vespro;\n\n')
        
        # Export forms table
        f.write('-- Forms table data\n')
        cur.execute('SELECT * FROM vespro.forms ORDER BY created_at')
        forms = cur.fetchall()
        
        for form in forms:
            columns = ', '.join(form.keys())
            values = []
            for key, val in form.items():
                if key in ['calculated_values', 'metadata', 'extra']:
                    if val:
                        values.append(f"'{json.dumps(val)}'::jsonb")
                    else:
                        values.append("'{}'::jsonb")
                elif val is None:
                    values.append('NULL')
                elif isinstance(val, (int, float)):
                    values.append(str(val))
                else:
                    values.append(escape_string(val))
            
            f.write(f"INSERT INTO vespro.forms ({columns}) VALUES ({', '.join(values)});\n")
        
        print(f"Forms exported: {len(forms)} records")
        
        # Export cost_items table
        f.write('\n-- Cost_items table data\n')
        cur.execute('SELECT * FROM vespro.cost_items ORDER BY item_id')
        items = cur.fetchall()
        
        for item in items:
            columns = ', '.join(item.keys())
            values = []
            for key, val in item.items():
                if key in ['extra']:
                    if val:
                        values.append(f"'{json.dumps(val)}'::jsonb")
                    else:
                        values.append("'{}'::jsonb")
                elif val is None:
                    values.append('NULL')
                elif isinstance(val, bool):
                    values.append('TRUE' if val else 'FALSE')
                elif isinstance(val, (int, float)):
                    values.append(str(val))
                else:
                    values.append(escape_string(val))
            
            f.write(f"INSERT INTO vespro.cost_items ({columns}) VALUES ({', '.join(values)});\n")
        
        print(f"Cost_items exported: {len(items)} records")
        
        # Export materials table (if any)
        cur.execute('SELECT * FROM vespro.materials')
        materials = cur.fetchall()
        if materials:
            f.write('\n-- Materials table data\n')
            for material in materials:
                columns = ', '.join(material.keys())
                values = []
                for key, val in material.items():
                    if val is None:
                        values.append('NULL')
                    elif isinstance(val, (int, float)):
                        values.append(str(val))
                    else:
                        values.append(escape_string(val))
                f.write(f"INSERT INTO vespro.materials ({columns}) VALUES ({', '.join(values)});\n")
            print(f"Materials exported: {len(materials)} records")
        
        # Export cost_groups table (if any)
        cur.execute('SELECT * FROM vespro.cost_groups')
        groups = cur.fetchall()
        if groups:
            f.write('\n-- Cost_groups table data\n')
            for group in groups:
                columns = ', '.join(group.keys())
                values = []
                for key, val in group.items():
                    if val is None:
                        values.append('NULL')
                    elif isinstance(val, (int, float)):
                        values.append(str(val))
                    else:
                        values.append(escape_string(val))
                f.write(f"INSERT INTO vespro.cost_groups ({columns}) VALUES ({', '.join(values)});\n")
            print(f"Cost_groups exported: {len(groups)} records")
    
    cur.close()
    conn.close()
    print("\nExport completed successfully!")
    print(f"File size: {os.path.getsize('insertdata.sql')} bytes")

if __name__ == '__main__':
    export_to_sql()